spec/models/notification_spec.rb

```diff
@@ -140,6 +140,66 @@
     end
   end
 
+  describe '.paginate_groups_by_max_id' do
+    let(:account) { Fabricate(:account) }
+
+    let!(:notifications) do
+      ['group-1', 'group-1', nil, 'group-2', nil, 'group-1', 'group-2', 'group-1']
+        .map { |group_key| Fabricate(:notification, account: account, group_key: group_key) }
+    end
+
+    context 'without since_id or max_id' do
+      it 'returns the most recent notifications, only keeping one notification per group' do
+        expect(described_class.without_suspended.paginate_groups_by_max_id(4).pluck(:id))
+          .to eq [notifications[7].id, notifications[6].id, notifications[4].id, notifications[2].id]
```

Comment:
This is sort of hard to follow in terms of which array position maps back to which notification in the created set ... might benefit from per-record better named vars?

Reply:
yeah, that would make sense

Reply:
I tried doing that, using multiple `let!` to enforce the correct order, with objects named `group1_notif1`, `group1_notif2`, `ungrouped_notif1` and so on, which does improve readability but contradicts multiple of the rubocop rules. I'm not sure which is preferable, so postponing this.